cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/WorkflowStatsSlaHttpHandler.java

```diff
@@ -0,0 +1,248 @@
+      for (int j = percentileStart; j < count; j++) {
+        percentileRun.add(workflowRunRecords.get(j).getWorkflowRunId());
+      }
+      percentileToRunids.put(Double.toString(i), percentileRun);
+      percentileToTime.put(Double.toString(i), workflowRunRecords.get(percentileStart).getTimeTaken());
+    }
+
+    Map<String, List<Long>> actionToRunRecord = getActionRuns(workflowRunRecords);
+
+    Map<String, Map<String, Long>> actionToPercentile = Maps.newHashMap();
+    for (Map.Entry<String, List<Long>> entry : actionToRunRecord.entrySet()) {
+      long mean = 0;
+      for (long iterator: entry.getValue()) {
+        mean += iterator;
+      }
+      mean /= entry.getValue().size();
+      Map temp = Maps.newHashMap();
+      temp.put("mean", mean);
+      actionToPercentile.put(entry.getKey(), temp);
+    }
+
+    for (Map.Entry<String, List<Long>> entry : actionToRunRecord.entrySet()) {
+      List<Long> runList = entry.getValue();
+      Collections.sort(runList);
+      for (double percentile : percentiles) {
+        long percentileValue = runList.get((int) ((percentile * runList.size()) / 100));
+        actionToPercentile.get(entry.getKey()).put(Double.toString(percentile), percentileValue);
+      }
+    }
+    
+    BasicStatistics basicStatistics = new BasicStatistics(startTime, endTime, count, average, percentileToTime,
+                                                          percentileToRunids, actionToPercentile);
+
+    responder.sendJson(HttpResponseStatus.OK, basicStatistics);
+  }
+
+  private List<WorkflowDataset.WorkflowRunRecord> sort(List<WorkflowDataset.WorkflowRunRecord> workflowRunRecords) {
+    Collections.sort(workflowRunRecords, new Comparator<WorkflowDataset.WorkflowRunRecord>() {
+      @Override
+      public int compare(WorkflowDataset.WorkflowRunRecord o1, WorkflowDataset.WorkflowRunRecord o2) {
+        if (o1.getTimeTaken() > o2.getTimeTaken()) {
+          return 1;
+        } else if (o1.getTimeTaken() < o2.getTimeTaken()) {
+          return -1;
+        } else {
+          return 0;
+        }
+      }
+    });
+    return workflowRunRecords;
+  }
+
+  private Map<String, List<Long>> getActionRuns(List<WorkflowDataset.WorkflowRunRecord> workflowRunRecords) {
+    Map<String, List<Long>> actionToRunRecord = Maps.newHashMap();
+    for (WorkflowDataset.WorkflowRunRecord workflowRunRecord: workflowRunRecords) {
+      for (WorkflowDataset.ActionRuns runs: workflowRunRecord.getActionRuns()) {
+        List<Long> runList;
+        if (actionToRunRecord.get(runs.getName()) == null) {
+          runList = Lists.newArrayList();
+        } else {
+          runList = actionToRunRecord.get(runs.getName());
+        }
+        runList.add(runs.getTimeTaken());
+        actionToRunRecord.put(runs.getName(), runList);
+      }
+    }
+    return actionToRunRecord;
+  }
+
+  @VisibleForTesting
+  /**
+   * Internal class to create Statistics
+   */
+  public static class BasicStatistics {
```

Comment:
For visible for testing only, prefer to be non-public, but package-private instead.